instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/config/Config.java

```diff
@@ -54,189 +42,63 @@
+  private static volatile Config INSTANCE = DEFAULT;
 
-    sqlNormalizerEnabled =
-        getBooleanSettingFromEnvironment(SQL_NORMALIZER_ENABLED, DEFAULT_SQL_NORMALIZER_ENABLED);
-
-    kafkaClientPropagationEnabled =
-        getBooleanSettingFromEnvironment(
-            KAFKA_CLIENT_PROPAGATION_ENABLED, DEFAULT_KAFKA_CLIENT_PROPAGATION_ENABLED);
-
-    hystrixTagsEnabled =
-        getBooleanSettingFromEnvironment(HYSTRIX_TAGS_ENABLED, DEFAULT_HYSTRIX_TAGS_ENABLED);
-
-    endpointPeerServiceMapping = getMapSettingFromEnvironment(ENDPOINT_PEER_SERVICE_MAPPING);
-
-    log.debug("New instance: {}", this);
+  /**
+   * Sets the agent configuration singleton. This method is only supposed to be called once, from
+   * the agent classloader just before the first instrumentation is loaded (and before {@link
+   * Config#get()} is used for the first time).
+   */
+  public static void internalInitializeConfig(Config config) {
+    if (INSTANCE != DEFAULT) {
+      log.warn("Config#INSTANCE was already set earlier");
+      return;
+    }
+    INSTANCE = requireNonNull(config);
   }
 
-  // Read order: Properties -> Parent
-  private Config(Properties properties, Config parent) {
-    exporterJar = properties.getProperty(EXPORTER_JAR, parent.exporterJar);
-    exporter = properties.getProperty(EXPORTER, parent.exporter);
-
-    propagators = getPropertyListValue(properties, PROPAGATORS, parent.propagators);
-
-    traceEnabled = getPropertyBooleanValue(properties, TRACE_ENABLED, parent.traceEnabled);
-    integrationsEnabled =
-        getPropertyBooleanValue(properties, INTEGRATIONS_ENABLED, parent.integrationsEnabled);
-
-    excludedClasses =
-        getPropertyListValue(properties, TRACE_CLASSES_EXCLUDE, parent.excludedClasses);
-
-    scopeDepthLimit =
-        getPropertyIntegerValue(properties, SCOPE_DEPTH_LIMIT, parent.scopeDepthLimit);
-
-    runtimeContextFieldInjection =
-        getPropertyBooleanValue(
-            properties, RUNTIME_CONTEXT_FIELD_INJECTION, parent.runtimeContextFieldInjection);
-
-    traceAnnotations = properties.getProperty(TRACE_ANNOTATIONS, parent.traceAnnotations);
-
-    traceMethods = properties.getProperty(TRACE_METHODS, parent.traceMethods);
-    traceAnnotatedMethodsExclude =
-        properties.getProperty(
-            TRACE_ANNOTATED_METHODS_EXCLUDE, parent.traceAnnotatedMethodsExclude);
-
-    traceExecutorsAll =
-        getPropertyBooleanValue(properties, TRACE_EXECUTORS_ALL, parent.traceExecutorsAll);
-    traceExecutors = getPropertyListValue(properties, TRACE_EXECUTORS, parent.traceExecutors);
-
-    sqlNormalizerEnabled =
-        getPropertyBooleanValue(properties, SQL_NORMALIZER_ENABLED, parent.sqlNormalizerEnabled);
-
-    kafkaClientPropagationEnabled =
-        getPropertyBooleanValue(
-            properties, KAFKA_CLIENT_PROPAGATION_ENABLED, parent.kafkaClientPropagationEnabled);
+  public static Config get() {
```

Comment:
Might add a recommendation to store config values as constants if they're referenced in performance-sensitive code.